arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/raise/Fold.kt

```diff
@@ -219,24 +220,37 @@ public inline fun <Error, A, B> foldUnsafe(
  * but **this only occurs** when composing `traced`.
  * The stacktrace creation is disabled if no `traced` calls are made within the function composition.
  */
+@OptIn(DelicateRaiseApi::class)
 @ExperimentalTraceApi
 public inline fun <Error, A> Raise<Error>.traced(
   @BuilderInference block: Raise<Error>.() -> A,
   trace: (trace: Trace, error: Error) -> Unit
 ): A {
-  val isOuterTraced = this is DefaultRaise && isTraced
-  val nested: DefaultRaise = if (isOuterTraced) this as DefaultRaise else DefaultRaise(true)
+  val nested = DefaultRaise(true)
   return try {
-    block.invoke(nested)
-  } catch (e: CancellationException) {
+    block(nested).also { nested.complete() }
+  } catch (e: Traced) {
+    nested.complete()
     val r: Error = e.raisedOrRethrow(nested)
     trace(Trace(e), r)
-    if (isOuterTraced) throw e else raise(r)
+    // If our outer Raise happens to be traced
+    // Then we want the stack trace to match the inner one
+    try {
+      raise(r)
+    } catch (rethrown: Traced) {
+      throw rethrown.withCause(e)
```

Comment:
```suggestion
      throw Traced(raised, raise, cause)
```

Reply:
Oh... right 😓 I missed that. No, this is much better! You definitely made the right call on that one 😉
Great work @kyay10! 👏 👏 👏